src/libraries/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs

```diff
@@ -283,23 +280,22 @@ public override bool Equals(object obj)
         private sealed class InstanceRestriction : BindingRestrictions
         {
             private readonly Expression _expression;
-            private readonly object _instance;
+            private readonly object? _instance;
 
-            internal InstanceRestriction(Expression parameter, object instance)
+            internal InstanceRestriction(Expression parameter, object? instance)
             {
                 Debug.Assert(parameter != null);
                 _expression = parameter;
                 _instance = instance;
             }
 
-            public override bool Equals(object obj)
+            public override bool Equals(object? obj)
             {
-                var other = obj as InstanceRestriction;
-                return other?._expression == _expression && other._instance == _instance;
+                return obj is InstanceRestriction other && other._expression == _expression && other._instance == _instance;
             }
 
             public override int GetHashCode()
-                => InstanceRestrictionHash ^ RuntimeHelpers.GetHashCode(_instance) ^ _expression.GetHashCode();
+                => InstanceRestrictionHash ^ RuntimeHelpers.GetHashCode(_instance!) ^ _expression.GetHashCode();
```

Comment:
>_instance [](start = 72, length = 9)

Does `RuntimeHelpers.GetHashCode()` require the argument to be non-null? Can we get here with `_instance == null`?

Reply:
> Does RuntimeHelpers.GetHashCode() require the argument to be non-null? 

By the existing annotations yes it is require non null
> Can we get here with _instance == null?

Yes, actually the only call to the conrstuctor called with null. https://github.com/dotnet/runtime/blob/7d4888b4fec2d66206585acc92c8f601007970bf/src/libraries/System.Linq.Expressions/src/System/Dynamic/BindingRestrictions.cs#L82